src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs

```diff
@@ -7612,5 +7612,477 @@ .maxstack  1
+                int V_3, //t
+                T V_4,
+                int? V_5,
+                int V_6)
+  IL_0000:  nop
+  IL_0001:  ldarg.0
+  IL_0002:  stloc.2
+  IL_0003:  ldloc.2
+  IL_0004:  stloc.s    V_4
+  IL_0006:  ldloc.s    V_4
+  IL_0008:  stloc.0
+  IL_0009:  ldloc.0
+  IL_000a:  box        ""T""
+  IL_000f:  brtrue.s   IL_0013
+  IL_0011:  br.s       IL_0038
+  IL_0013:  ldloc.0
+  IL_0014:  box        ""T""
+  IL_0019:  isinst     ""int?""
+  IL_001e:  unbox.any  ""int?""
+  IL_0023:  stloc.s    V_5
+  IL_0025:  ldloca.s   V_5
+  IL_0027:  call       ""int int?.GetValueOrDefault()""
+  IL_002c:  stloc.1
+  IL_002d:  ldloca.s   V_5
+  IL_002f:  call       ""bool int?.HasValue.get""
+  IL_0034:  brfalse.s  IL_0038
+  IL_0036:  br.s       IL_003a
+  IL_0038:  br.s       IL_0043
+  IL_003a:  ldloc.1
+  IL_003b:  stloc.3
+  IL_003c:  br.s       IL_003e
+  IL_003e:  ldloc.3
+  IL_003f:  stloc.s    V_6
+  IL_0041:  br.s       IL_0048
+  IL_0043:  ldc.i4.0
+  IL_0044:  stloc.s    V_6
+  IL_0046:  br.s       IL_0048
+  IL_0048:  ldloc.s    V_6
+  IL_004a:  ret
+}"
+            );
+        }
+
+        [Fact, WorkItem(16195, "https://github.com/dotnet/roslyn/issues/16195")]
+        public void TestMatchWithTypeParameter_03()
+        {
+            var source =
+@"using System;
+class Program
+{
+    public static void Main(string[] args)
+    {
+        Console.Write(M1<int>(2));
+        Console.Write(M2<int>(3));
+        Console.Write(M1<int>(1.1));
+        Console.Write(M2<int>(1.1));
+    }
+    public static T M1<T>(ValueType o) where T : struct
+    {
+        return o is T t ? t : default(T);
+    }
+    public static T M2<T>(ValueType o) where T : struct
+    {
+        switch (o)
+        {
+            case T t:
+                return t;
+            default:
+                return default(T);
+        }
+    }
+}
+";
+            var compVerifier = CompileAndVerify(source,
+                options: TestOptions.DebugDll.WithOutputKind(OutputKind.ConsoleApplication),
+                expectedOutput: "2300");
+        }
+
+        [Fact, WorkItem(16195, "https://github.com/dotnet/roslyn/issues/16195")]
+        public void TestMatchWithTypeParameter_04()
+        {
+            var source =
+@"using System;
+class Program
+{
+    public static void Main(string[] args)
+    {
+        var x = new X();
+        Console.Write(M1<B>(new A()) ?? x);
+        Console.Write(M2<B>(new A()) ?? x);
+        Console.Write(M1<B>(new B()) ?? x);
+        Console.Write(M2<B>(new B()) ?? x);
+    }
+    public static T M1<T>(A o) where T : class
+    {
+        return o is T t ? t : default(T);
+    }
+    public static T M2<T>(A o) where T : class
+    {
+        switch (o)
+        {
+            case T t:
+                return t;
+            default:
+                return default(T);
+        }
+    }
+}
+class A
+{
+}
+class B : A
+{
+}
+class X : B { }
+";
+            CreateStandardCompilation(source).VerifyDiagnostics(
```

Comment:
>CreateStandardCompilation(source).VerifyDiagnostics( [](start = 12, length = 52)

Consider explicitly using C# 7 version, otherwise this test will break once we bump the default version.